File: src/primitives/Address/toShortHex.test.ts
import { describe, expect, it } from "vitest";
import { Address } from "./Address.js";
import { toShortHex } from "./toShortHex.js";

describe("toShortHex", () => {
	it("shortens address with default lengths", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr);
		expect(short).toBe("0x742d35...51e3");
	});

	it("uses custom prefix length", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr, 8, 4);
		expect(short).toBe("0x742d35cc...51e3");
	});

	it("uses custom suffix length", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr, 6, 6);
		expect(short).toBe("0x742d35...251e3");
	});

	it("returns full hex if lengths exceed address length", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr, 30, 30);
		expect(short).toBe("0x742d35cc6634c0532925a3b844bc9e7595f251e3");
	});

	it("returns full hex if prefix + suffix >= 40", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr, 20, 20);
		expect(short).toBe("0x742d35cc6634c0532925a3b844bc9e7595f251e3");
	});

	it("handles zero address", () => {
		const addr = Address.zero();
		const short = toShortHex(addr);
		expect(short).toBe("0x000000...0000");
	});

	it("works with Address namespace method", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = Address.toShortHex(addr);
		expect(short).toBe("0x742d35...51e3");
	});

	it("works with instance method", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = addr.toShortHex();
		expect(short).toBe("0x742d35...51e3");
	});

	it("includes ellipsis separator", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr);
		expect(short).toContain("...");
	});

	it("preserves 0x prefix", () => {
		const addr = Address.fromHex(
			"0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		);
		const short = toShortHex(addr);
		expect(short.startsWith("0x")).toBe(true);
	});
});

File: src/primitives/Address/toShortHex_test.go
package address

import (
	"testing"
)

func TestToShortHex(t *testing.T) {
	addr, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")

	tests := []struct {
		name      string
		prefixLen int
		suffixLen int
		want      string
	}{
		{
			name:      "default lengths",
			prefixLen: 6,
			suffixLen: 4,
			want:      "0x742d35...51e3",
		},
		{
			name:      "custom lengths",
			prefixLen: 8,
			suffixLen: 6,
			want:      "0x742d35cc...f251e3",
		},
		{
			name:      "full address",
			prefixLen: 40,
			suffixLen: 0,
			want:      "0x742d35cc6634c0532925a3b844bc9e7595f251e3",
		},
		{
			name:      "very short prefix",
			prefixLen: 2,
			suffixLen: 2,
			want:      "0x74...e3",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := addr.ToShortHex(tt.prefixLen, tt.suffixLen)
			if got != tt.want {
				t.Errorf("ToShortHex() = %s, want %s", got, tt.want)
			}
		})
	}
}

func TestFormat(t *testing.T) {
	addr, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	formatted := addr.Format()

	want := "0x742d35Cc6634c0532925a3b844bc9e7595F251E3"
	if formatted != want {
		t.Errorf("Format() = %s, want %s", formatted, want)
	}
}
